lib/init.php

```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
 	);
 }
 add_action( 'admin_menu', 'gutenberg_menu', 9 );
+
+if ( ! function_exists( 'html_contains_block' ) ) {
+	/**
+	 * Recursively find a block by attribute.
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name The block type to search for.
+	 * @param  string  $attribute_name The attribute name to search for.
+	 * @param  string  $attribute_value The attribute value to search for.
+	 *
+	 * @return bool    True if block is found, false otherwise
+	 */
+	function html_contains_block( $html, $block_name, $attribute_name = null, $attribute_value = null ) {
+		$at = 0;
+
+		/**
+		 * This is the same regex as the one used in the block parser.
+		 * It is better to use this solution to look for a block's existence
+		 * in a document compared to having to parsing the blocks in the
+		 * document, avoiding all the performance drawbacks of achieving
+		 * a full representation of block content just to check if one block
+		 * is there.
```

Comment:
this is a great comment! I would consider adding a link of some form (doesn't need to be a _link_ in fact, but some way to point people to the right place).

```php
/**
 * …
 *
 * @see WP_Block_Parser
 *
 * @var string
 */
```